### PR TITLE
Add support for Laravel 9

### PR DESCRIPTION
**Pull request by Nielsvanpach**
I don't think it' possible to add Laravel 9 support without dropping 7.3 and 7.4 support.

**Current issues**

- [Codeception](https://github.com/Codeception/Codeception) v4 does not support symfony/finder v6 but they are working on a v5 release with support. 
symfony/finder v6 is used in Laravel 9.
- Codeception v5 has a conflicting package constraint for sebastian/diff: https://github.com/Codeception/Codeception/blob/5.0/composer.json#L37
https://github.com/vimeo/psalm/blob/4.18.x/composer.json#L36
- [weirdan/codeception-psalm-module](https://github.com/psalm/codeception-psalm-module) does not support Codeception v5 yet. We could try to open a PR to that repo.

Hopefully if this is fixed, all the dependency issues are solved and we can work on proper Laravel 9 compatibility.

### CODE REVIEW

File: composer.json
Comment (mr-feek) [inline composer.json:16]: i think we should still support laravel 8, still has over a year of bug fix support from laravel. Shouldn't this then be `"illuminate/config": "^6.0 || ^8.0 || ^9.0",`?

File: .github/workflows/test.yml
Comment (caugner) [inline .github/workflows/test.yml]: No need to support `7.*`, that one is no longer supported.

File: composer.json
Comment (caugner) [inline composer.json]: This should include `^8.0`, here and below.

```suggestion
        "illuminate/config": "^6.0 || ^8.0 || ^9.0",
```

File: .github/workflows/test.yml
Comment (caugner) [inline .github/workflows/test.yml]: Won't this ignore the `illuminate_version` above and just install the lowest `6.x` for `--prefer-lowest` and the highest `9.x` for `--prefer-stable`?

File: .github/workflows/test.yml
Comment (caugner) [inline .github/workflows/test.yml:13]: `--prefer-stable` is not the opposite of `--prefer-lowest`, it also makes sense in combination.

Actually, I think that `--prefer-stable` is even the default, but I'm not totally sure.

**Review by caugner — CHANGES_REQUESTED** [review]
Good initiative, but I would suggest only adding `9.x` to the `illuminate_version` matrix for now.

Previously, the tests were testing the lowest and highest version of each Laravel major version, but I don't think the status quo of this PR still does that.

File: composer.json
Comment (zlodes) [reply composer.json:16]: @mr-feek laravel added native template-params for Collections. 
I think it will be easier to split two tags: `^1.0` for laravel 6 | 8 and `^2.0` for `^9.0`.

I want to believe that this plugin will be unnecessary in the nearest future because of full native support (e.g. Builder).

**Comment by zlodes** [comment]
I think we need to remove restrictions of `barryvdh/laravel-ide-helper`, because these versions don't require `laravel/framework: ^9.0`

File: composer.json
Comment (caugner) [reply composer.json:16]: > I want to believe that this plugin will be unnecessary in the nearest future because of full native support (e.g. Builder).

Unfortunately, a dedicated plugin will be necessary as long as Laravel does not have conditional type annotations for "overloaded" functions like `str()` (new) and `Request::validated()` (changed in L9), which return a different type based on the number of parameters provided.

**Comment by zlodes** [comment]
@caugner Larastan now (^2.0) has several nasty bugs with generics. :|

We use phpstan and psalm together because it increases global code quality. Nowadays the functionality difference between `psalm` and `phpstan` is very huge, but it's not bad, v.v. it's cool that these tools are improving.

**Comment by canvural** [comment]
> @caugner Larastan now (^2.0) has several nasty bugs with generics. :|

What are those nasty bugs? Maybe we can fix them if you open an issue 😊

**Comment by zlodes** [comment]
@canvural sorry. I meant phpstan issues, e.g.: https://github.com/phpstan/phpstan/issues/6505

And if I update Larastan to `^2.0` something around generics will broken.
We use Psalm and PhpStan together and now it seems that we need to release `psalm-plugin-laravel` because of many changes in Laravel 9.0. And then it will be possible to use Larastan ^2.0 and Psalm.

**Comment by HDVinnie** [comment]
Any timeline on when Laravel 9 will be supported?

**Comment by mr-feek** [comment]
@HDVinnie if a PR is made with green CI, I'll happily merge! I don't have the time to figure it out myself right now

File: composer.json
Comment (Nielsvanpach) [reply composer.json]: right!

File: .github/workflows/test.yml
Comment (Nielsvanpach) [reply .github/workflows/test.yml:13]: Indeed, both can be used in combination I think. `prefer-stable` will check for packages with `minimum-stability` set to stable. But I don't think it's the default (and if so, it's still explicit this way). It's just an extra check to see if this package is easily installable, feel free to improve if possible!

**Comment by tm1000** [comment]
> Due to limited resources, I will probably have to migrate at least one of my Laravel projects to larastan (maintained by Nuno who _works_ for Laravel), because it requires Laravel 9 asap.

This is really unfortunate. You are a maintainer of this project and yet you are essentially abandoning it and moving to phpstan because you don't have the time to maintain it.

To me the outcome is obvious. If you want to use Laravel 9 then do not use Psalm, use phpstan. That is unfortunate.

**Comment by Nielsvanpach** [comment]
So basically you are disappointed that someone no longer has (unpaid) time to support an open source project and you feel the need to let that person know that it shocks you? 🙃

**Comment by tm1000** [comment]
No. It's shocking that one of the maintainers said he has no time to work on this so he's going to use another project instead. That should not have been said. Anyone who comes here for help will see that and think to themselves why use this"

There have been several proposed fixes to support new ide helper and all have been rejected because of the continued need to support laravel 6.0 even though one could just do a sem ver release of a new major version. 

The fork that exists is also an issue because it violates packagist.org guidelines which state that one should not publish packages that are forks. Yet it exists in packagist.org and additionally it's listed on the psalm site. The reasoning the fork exists in the first place is because this project can't find a solution for a long standing issue in regards to ide helper. That's a big big issue in itself.

**Comment by canvural** [comment]
I don't think @caugner is a maintainer of this project. He just contributes a lot. So it's totally fine if he says he will not work on this.

**Comment by tm1000** [comment]
@caugner and then yes. I am totally wrong. I will accept that.

**Comment by xrobau** [comment]
> So basically you are disappointed that someone no longer has (unpaid) time to support an open source project you rely on and you feel the need to let that person know that it shocks you? 🙃

Just FYI, @tm1000 was for a VERY long time the lead developer of FreePBX - the world's most popular open source PBX (also written in mainly PHP). He is well aware of how hard it is to try to 'Open Source' *and* make everyone happy. It is not easy, and it's also extremely easy to slightly mis-read something and end up with the totally wrong idea of what someone was trying to say.

I have no new suggestions on how to solve this common problem of, basically, people just losing interest in their projects! I just wanted to make sure that everyone knew that @tm1000 wasn't just blowing hot air, he actually knows what he's talking about (even if he was slightly off target!)

**Comment by caugner** [comment]
As @canvural has rightfully pointed out, I'm a *contributor* to this project, and I have never considered myself a maintainer of this project.. But then again, I do regret my message above, because it may have been discouraging to other fellow contributors (and the real maintainer, @mr-feek), and it obviously mislead others.

Apologies for my part, and now let's try to focus the discussion on *adding support for Laravel 9*, rather than talking about myself (or Larastan).

**Comment by mr-feek** [comment]
Always happy to add more maintainers after a proven track record of contributions. I try to merge every PR that has passing CI. I haven't worked with laravel too much for the last year or two due to a shift in focus at work, so I haven't had the personal drive to dedicate hours / week to this project for new features. 

At this point in time, I think I'm ok with dropping laravel 6 active maintenance (minus bug fixes) if that helps us to support laravel 9

**Comment by tm1000** [comment]
Are you ok with upgrading ide helper as well? I know the thought process was to not use ide helper and write your own but if you both don't have time then probably upgrading ide helper now is a good solution.

**Comment by HDVinnie** [comment]
@tm1000 This is FOSS....you cannot expect it to be maintained forever not matter how many ppl use it. Truthfully, you should never depend on a package your not willing to maintain. They should considers time savers but should still be reviewed like they are a PR. Pulling them from a private fork also allows you to merge from upstream and review the code for several things. Look at all the NPM drama lately....your coming off as ungrateful and as if the maintainers of this repo own you something when they dont owe you a thing.

**Comment by tm1000** [comment]
@HDVinnie Thanks for your comments. I think we've all moved past this. I already apologized. The discussion on that ended a while ago.

**Comment by HDVinnie** [comment]
> @HDVinnie Thanks for your comments. I think we've all moved past this. I already apologized. I encourage you to reread the thread in its entirety. The discussion on that ended a while ago.

Awhile ago lmao....just 6 hours ago you apologized. It doesn't matter if the guy was a maintainer or not. Never expect your owed updates on a FOSS package, thats my point. Anyway moving right along.....

**Comment by tm1000** [comment]
FYI Codeception v5 is set to require PHPUnit 10. Is that going to be an issue? This is why many things in composer.json for Codeception are marked as master-dev

https://github.com/Codeception/Codeception/commit/f909c6d8523bab3fc1e97b2f23a1b0e428121e7b

**Comment by mzur** [comment]
I tried to see what's blocking this PR (creating #221 and #223 in the process). So the main issue is Codeception v5 which introduces the following conflicts that were partly described above:

1. `weirdan/codeception-psalm-module` requires Codeception v4. I tried to update it to 5.0.0-alpha2 but I couldn't resolve the conflict caused by `codeception/module-filesystem:^2.0`.

2. `vimeo/psalm` requires `sebastian/diff: ^3.0 || ^4.0` which conflicts with Codeception v5 (requiring `dev-master`).

So in summary, this is blocked until Codeception v5 is properly released and its version constraints finalized. Codeception v5 [waits for the release of PHPUnit 10](https://github.com/Codeception/Codeception/discussions/6327) (which is already a year late with no new release date I could find).

After that, `weirdan/codeception-psalm-module` and possibly  `vimeo/psalm` need updates to support the new versions. Then this PR can proceed. Please correct me if I'm wrong.

**Comment by tm1000** [comment]
@mzur Yes that is the conclusion I got as well. This depends on PHPUnit 10. Then it will depend on Psalm itself supporting PHPUnit 10.

**Comment by jrmajor** [comment]
> `vimeo/psalm` requires `sebastian/diff: ^3.0 || ^4.0` which conflicts with Codeception v5.

@mzur @Nielsvanpach Psalm works fine with PHPUnit 10 for me with `"sebastian/diff": "5.0.x-dev as 4.0.4"`. Maybe the same trick would work there?

**Comment by mzur** [comment]
Seems like Codeception v5 [will still support PHPUnit 9](https://github.com/Codeception/Codeception/pull/6380) for now. This should make things easier.

**Comment by Nielsvanpach** [comment]
Solved the dependency issues and added the sole class of [weirdan/codeception-psalm-module](https://github.com/weirdan/codeception-psalm-module) to this repo for now (Module.php), to bypass the composer restrictions. But it seems like the package is not yet compatible with Codeception v5, tests are failing: https://github.com/psalm/psalm-plugin-laravel/actions/runs/2014834434

**Comment by tm1000** [comment]
@Nielsvanpach I issued a PR against your `laravel-9` branch that fixes the issues you noticed (see: https://github.com/Nielsvanpach/psalm-plugin-laravel/pull/1)

Now the tests run, some of the tests do fail but now we can go through and fix each test.

**Comment by Nielsvanpach** [comment]
> @Nielsvanpach I issued a PR against your `laravel-9` branch that fixes the issues you noticed (see: [Nielsvanpach#1](https://github.com/Nielsvanpach/psalm-plugin-laravel/pull/1))

Great thanks!

**Comment by tm1000** [comment]
@Nielsvanpach Ok this PR (https://github.com/Nielsvanpach/psalm-plugin-laravel/pull/2) fixes the tests (except for 1 or 2....maybe?) and also fixes the merge conflict with master

**Comment by tm1000** [comment]
Ok so this should be a little easier for someone to fix (if not me).

The failing tests are due to php 8.1 incpompatibilies when using prefer-lowest because the issues were made in as a non-breaking change in the same release.

Additionally in Laravel 9 more specific models are returned rather than just Model itself so the tests need to be adjusted to reflect that for Laravel 9 as opposed to 8

**Review by mzur — COMMENTED** [review]
Is there any specific reason why the tests for `firstOrCreate()`, `firstOrNew()` and `factory()` were removed? These can still be used in Laravel 9.

**Comment by mzur** [comment]
I created a [PR](https://github.com/Nielsvanpach/psalm-plugin-laravel/pull/3) to this branch that should fix the remaining failing tests. Does anyone have a better idea to fix the PHP 8.1 issues than to `require-dev` working versions of the incompatible packages?

**Comment by tm1000** [comment]
Thanks @mzur !

**Comment by mzur** [comment]
All checks are green now so this may be ready to merge?

**Comment by tm1000** [comment]
@mzur thanks for getting it the final way there!

**Review by tm1000 — APPROVED** [review]

File: tests/Support/Module.php
Comment (mr-feek) [inline tests/Support/Module.php:23]: Hey could you help me understand why we need this duplicated? Do we need some changes merged upstream to `codeception-psalm-module`? @weirdan has been great at merging prs fast! Let me know

File: .github/workflows/test.yml
Comment (mr-feek) [inline .github/workflows/test.yml:12]: It would be great if we could support all versions in the same major release (1.x) -- Could you help me understand what might be a breaking change not allowing us to support these still? (I want to continue supporting laravel 6 til EOL, which is in september).

Comment (tm1000) [reply .github/workflows/test.yml:12]: > I don't think it' possible to add Laravel 9 support without dropping 7.3 and 7.4 support.

Laravel 6 won't work well on php 8. It's certainly not supported. That's why I believe laravel 6 support was dropped. 

Is there a major issue with just issuing a 2.0 release for this PR? I don't understand the hesitation at doing a split release as has been already mentioned in this PR previously. See: https://github.com/psalm/psalm-plugin-laravel/pull/213#discussion_r803103993

Comment (jrmajor) [reply .github/workflows/test.yml:12]: Larastan (Laravel plugin for PHPStan) also released a new major version for Laravel 9.

> Main reason we have a major version increase is we dropped support for Laravel versions older than 9. Because now all the Collection stubs Larastan had are moved to the Laravel core!

Comment (mr-feek) [reply .github/workflows/test.yml:12]: My hesitation is just more maintenance overhead -- I would *think* we could solve that by adjusting the test matrix to not run laravel 6 on php 8 (like we are already doing with other tests) .

It's not a blocker, would just like to understand what exactly is causing the need for multiple versions

File: tests/Support/Module.php
Comment (tm1000) [reply tests/Support/Module.php:23]: It's duplicated because of the changes to codeception. Doing the PR against https://github.com/psalm/codeception-psalm-module would require a newer version of codeception which doesnt support PHP 7.x.

See: https://www.diffchecker.com/kbix1BTd

Specifically params that *were* strings are now `PyStringNode` *object* and additionally, exit codes are now integers instead of strings

Comment (tm1000) [reply tests/Support/Module.php:23]: I just went through https://github.com/psalm/codeception-psalm-module and unfortunately even checking it out and running the tests there without doing any code modifications there are *13* unit test failures.

The issue is with the deprecated usage of `totallyTyped` and yes usually Weirdan is quick on PR changes but @mzur already did a PR for this back in February which is still *not* merged. https://github.com/psalm/codeception-psalm-module/pull/39

File: composer.json
Comment (mr-feek) [inline composer.json:39]: curious what we need this for? I don't see it referenced in the code

File: composer.json
Comment (mr-feek) [inline composer.json]: this looks accidental

File: composer.json
Comment (tm1000) [reply composer.json:39]: I know that if `ramsey/collection` is less than 1.2.0 it blows up on PHP 8.1 tests because Return Type will Change errors.

Comment (mzur) [reply composer.json:39]: These were all the packages that didn't work with PHP 8.1 at `prefer-lowest`. I didn't know any better than to require the versions compatible with PHP 8.1 to fix the failing tests. The main issue is the breaking change of return type checks in the minor release of PHP and all these packages claiming to work with PHP `^8.0` although they are not.

**Comment by mr-feek** [comment]
Thanks for the PR! cutting a new major release, 2.0

**Pull request merged**
